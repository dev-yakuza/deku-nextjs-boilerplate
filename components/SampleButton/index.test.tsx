import { render, screen, fireEvent } from '@testing-library/react'
import { SampleButton } from './index'

const mockButton = jest.fn()

jest.mock('@mui/material', () => {
  const { Button, ...rest } = jest.requireActual('@mui/material')

  const MockButton = (props: unknown) => {
    mockButton(props)
    return <Button {...props} />
  }

  return {
    Button: MockButton,
    ...rest,
  }
})

describe('<SampleButton />', () => {
  beforeEach(() => {
    mockButton.mockClear()
  })

  it('Primary', async () => {
    const { container } = render(
      <SampleButton color="primary" label="Button" />,
    )

    expect(mockButton.mock.calls[0][0].variant).toBe('contained')
    expect(mockButton.mock.calls[0][0].color).toBe('primary')
    expect(mockButton.mock.calls[0][0].size).toBe('medium')
    expect(mockButton.mock.calls[0][0].backgroundColor).toBe(undefined)
    expect(mockButton.mock.calls[0][0].children).toBe('Button')

    expect(container).toMatchSnapshot()
  })

  it('backgroundColor', async () => {
    render(<SampleButton backgroundColor="red" label="Button" />)
    expect(screen.getByText('Button')).toHaveStyle('background-color: red')
  })

  it('size', async () => {
    const { rerender } = render(<SampleButton size="small" label="Button" />)
    expect(mockButton.mock.calls[0][0].size).toBe('small')

    rerender(<SampleButton size="medium" label="Button" />)
    expect(mockButton.mock.calls[1][0].size).toBe('medium')

    rerender(<SampleButton size="large" label="Button" />)
    expect(mockButton.mock.calls[2][0].size).toBe('large')
  })

  it('label', async () => {
    const { rerender } = render(<SampleButton label="Button" />)
    expect(screen.queryByText('Button')).toBeInTheDocument()
    expect(screen.queryByText('Test Button')).not.toBeInTheDocument()

    rerender(<SampleButton label="Test Button" />)
    expect(screen.queryByText('Button')).not.toBeInTheDocument()
    expect(screen.queryByText('Test Button')).toBeInTheDocument()
  })

  it('onClick', async () => {
    const onClick = jest.fn()
    render(<SampleButton label="Button" onClick={onClick} />)

    expect(onClick).not.toHaveBeenCalled()
    fireEvent.click(screen.getByText('Button'))
    expect(onClick).toHaveBeenCalled()
  })
})
