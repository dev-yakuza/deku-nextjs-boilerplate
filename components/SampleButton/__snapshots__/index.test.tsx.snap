// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SampleButton /> Primary 1`] = `
.emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1.75;
  letter-spacing: 0.02857em;
  text-transform: uppercase;
  min-width: 64px;
  padding: 6px 16px;
  border-radius: 4px;
  -webkit-transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  color: #fff;
  background-color: #1976d2;
  box-shadow: 0px 3px 1px -2px rgba(0,0,0,0.2),0px 2px 2px 0px rgba(0,0,0,0.14),0px 1px 5px 0px rgba(0,0,0,0.12);
}

.emotion-0::-moz-focus-inner {
  border-style: none;
}

.emotion-0.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-0 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-0:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: #1565c0;
  box-shadow: 0px 2px 4px -1px rgba(0,0,0,0.2),0px 4px 5px 0px rgba(0,0,0,0.14),0px 1px 10px 0px rgba(0,0,0,0.12);
}

@media (hover: none) {
  .emotion-0:hover {
    background-color: #1976d2;
  }
}

.emotion-0:active {
  box-shadow: 0px 5px 5px -3px rgba(0,0,0,0.2),0px 8px 10px 1px rgba(0,0,0,0.14),0px 3px 14px 2px rgba(0,0,0,0.12);
}

.emotion-0.Mui-focusVisible {
  box-shadow: 0px 3px 5px -1px rgba(0,0,0,0.2),0px 6px 10px 0px rgba(0,0,0,0.14),0px 1px 18px 0px rgba(0,0,0,0.12);
}

.emotion-0.Mui-disabled {
  color: rgba(0, 0, 0, 0.26);
  box-shadow: none;
  background-color: rgba(0, 0, 0, 0.12);
}

.emotion-1 {
  overflow: hidden;
  pointer-events: none;
  position: absolute;
  z-index: 0;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  border-radius: inherit;
}

<div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium emotion-0"
    tabindex="0"
    type="button"
  >
    Button
    <span
      class="MuiTouchRipple-root emotion-1"
    />
  </button>
</div>
`;
